### PR TITLE
Use LTIUser to detect the LMS we are currently using

### PR DESCRIPTION
**Pull request by marcospri**
Based on the Moodle work we have a need to deduplicate code around files/pages/course copy.

As a first step we are going to remove different way to detect which LMS is the current one and settle in just one method that will work in all situations.

## Testing

No behaviour changes with this PR, but sanity checking a few things.

- [Reconfigure a Canvas assignment](https://hypothesis.instructure.com/courses/125/assignments/5115/edit?name=Why+did+all+these+elephants+die%3F+&due_at=null&points_possible=100)
- Pick Canvas files
- [Launch a gradable D2L assignment](https://aunltd.brightspacedemo.com/d2l/le/content/6782/viewContent/3348/View)
- Grade a student

### CODE REVIEW

File: lms/product/factory.py [hunk 1]
```diff
@@ -11,44 +12,9 @@
...
-    family = _get_family(request, ai)
-    product = _PRODUCT_MAP.get(family, Product).from_request(
-        request, dict(ai.settings) if ai else {}
-    )
+    product = _PRODUCT_MAP.get(family, Product).from_request(request, dict(ai.settings))
     product.family = family
     return product
-
-
-def _get_family(
```
Comment (marcospri) [inline lms/product/factory.py:24]: This method infers the product based on a number of methods

File: lms/product/factory.py [hunk 2]
```diff
@@ -11,44 +12,9 @@
...
-
-    if request.matched_route.name.startswith("blackboard_api."):
-        return Product.Family.BLACKBOARD
-
-    # If we are in an API request, where we are forwarding the product type ourselves
-    if request.content_type == "application/json" and "lms" in request.json:
-        return Product.Family(request.json["lms"]["product"])
-
-    # In an LTI launch we'll use the parameters available to guess
-    if product_name := request.lti_params.get("tool_consumer_info_product_family_code"):
```
Comment (marcospri) [inline lms/product/factory.py:40]: This one is based on information available during launch.

File: lms/product/factory.py [hunk 3]
```diff
@@ -11,44 +12,9 @@
...
-        return Product.Family(product_name)
-
-    # If we don't get a hint from LTI check a canvas specific parameter
-    if "custom_canvas_course_id" in request.lti_params:
```
Comment (marcospri) [inline lms/product/factory.py:44]: This one is also available during launch.

File: lms/product/family.py [hunk 4]
```diff
@@ -16,3 +16,21 @@ class Family(str, Enum):
     @classmethod
     def _missing_(cls, _value):
         return cls.UNKNOWN
+
+    @classmethod
+    def from_launch(cls, lti_params: dict):
```
Comment (marcospri) [inline lms/product/family.py:21]: Moved the method here from product.factory.

Now the method is not about "guessing" the product with different methods. It only looks at information available during launch.

File: lms/resources/_js_config/__init__.py [hunk 5]
```diff
@@ -318,9 +318,6 @@ def add_deep_linking_api(self):
                 "content_item_return_url": self._request.lti_params[
                     "content_item_return_url"
                 ],
-                "lms": {
```
Comment (marcospri) [inline lms/resources/_js_config/__init__.py:321]: We don't need to forward this manually anymore.

I'll mention where this were used to add them to the testing instructions.

This one applies to deep linking.

File: lms/services/lti_user.py [hunk 6]
```diff
@@ -37,6 +38,7 @@ def from_lti_params(self, application_instance, lti_params) -> LTIUser:
             lti={
                 "course_id": lti_params.get("context_id"),
                 "assignment_id": lti_params.get("resource_link_id"),
+                "product_family": Family.from_launch(lti_params),
```
Comment (marcospri) [inline lms/services/lti_user.py:41]: While building a LTIUser for the first time (ie, from a launch, not one that we serelized before) we get the product from the LTI params.

File: lms/services/lti_user.py [hunk 7]
```diff
@@ -57,6 +59,7 @@ def serialize(lti_user: LTIUser) -> dict:
             "lti": {
                 "course_id": lti_user.lti.course_id,
                 "assignment_id": lti_user.lti.assignment_id,
+                "product_family": lti_user.lti.product_family,
```
Comment (marcospri) [inline lms/services/lti_user.py:62]: We include the string in the serialized versions of LTIUser.

File: lms/services/lti_user.py [hunk 8]
```diff
@@ -75,6 +78,7 @@ def deserialize(self, **kwargs: dict) -> LTIUser:
         lti = LTI(
             course_id=lti_data["course_id"],
             assignment_id=lti_data["assignment_id"],
+            product_family=lti_data["product_family"],
```
Comment (marcospri) [inline lms/services/lti_user.py:81]: And when building a LTIUser form a serialized one, just use the string available there.

File: lms/views/api/sync.py [hunk 9]
```diff
@@ -8,10 +7,6 @@
 
 
 class APISyncSchema(PyramidRequestSchema):
-    class LMS(Schema):
```
Comment (marcospri) [inline lms/views/api/sync.py:11]: This is no longer required, we are not sending this information from JSConfig anymore.

File: lms/product/factory.py [hunk 10]
```diff
@@ -11,44 +12,9 @@
 
 def get_product_from_request(request) -> Product:
     """Get the correct product object from the provided request."""
-    ai = request.lti_user.application_instance if request.lti_user else None
+    ai = request.lti_user.application_instance
+    family = Family(request.lti_user.lti.product_family)
```
Comment (acelaya) [inline lms/product/factory.py:16]: What would happen here if `request.lti_user` is not set? We don't need to check that anymore?

**Review by acelaya — APPROVED** [review]
Looks good, and works as expected. I just had a query about a `request.lti_user` check that has been removed.

File: lms/product/factory.py (hunk 10 above)
Comment (marcospri) [reply lms/product/factory.py:16]: I had the same question while refactoring this.

My understanding is that that check was superfluous. If we don't have a lti_user is either:

- It's an LTI launch and we are building the LTIUser at that moment. The code in this PR should build the LTIUser first, populate `product_family` and then arrive here.

- It's an unauthenticated request. Unless I'm missing something important we shouldn't care/query for product in those case. If we found a use case were that's necessary we should return UNKNOWN here but ideally we'd know which those cases are.

Comment (acelaya) [reply lms/product/factory.py:16]: Gotcha

**Pull request merged**